craft_store/creds.py

```diff
@@ -0,0 +1,153 @@
+        except json.JSONDecodeError:
+            return CandidModel(v=marshalled_creds)  # pyright: ignore
+
+        if "t" in creds:
+            return CandidModel(**creds)
+        return CandidModel(v=marshalled_creds)  # pyright: ignore
+
+
+def marshal_candid_credentials(candid_creds: str) -> str:
+    """Serialize Candid credentials for storage.
+
+    This function creates a string that contains the desired `candid_creds` but also
+    stores their "type", for unmarshalling later with `unmarshal_candid_credentials()`.
+
+    :param candid_creds: The actual Candid credentials.
+    :return: A payload string ready to be passed to Auth.set_credentials()
+    """
+    return CandidModel(v=candid_creds).marshal()  # pyright: ignore
+
+
+def unmarshal_candid_credentials(marshalled_creds: str) -> str:
+    """Deserialize previously stored Candid credentials.
+
+    This function also handles backwards-compatibility by supporting `marshalled_creds`
+    from before we stored the `token_type`; it is meant to be called with the
+    returned value from Auth.get_credentials().
+
+    :param marshalled_creds: The credentials retrieved from auth storage.
+    :return: The actual Candid credentials.
+    """
+    try:
+        return CandidModel.unmarshal(marshalled_creds).value
+    except pydantic.ValidationError as err:
+        raise errors.CredentialsNotParseable(
+            "Expected valid Candid credentials"
+        ) from err
+
+
+class UbuntuOneMacaroons(BaseModel):
+    """Model representation of the set of macaroons used in Ubuntu SSO."""
+
+    r: str
+    d: str
+
+    @property
+    def root(self) -> str:
+        """Get the root macaroon."""
+        return self.r
+
+    @property
+    def discharge(self):
+        """Get the discharge macaroon."""
+        return self.d
+
+    def with_discharge(self, discharge: str) -> "UbuntuOneMacaroons":
+        """Create a copy of this UbuntuOneMacaroons with a different discharge macaroon."""
+        return self.copy(update={"d": discharge})
+
+
+class UbuntuOneModel(BaseModel):
+    """Model for Ubuntu One credentials."""
+
+    t: Literal["u1-macaroon"] = Field("u1-macaroon")
+    v: UbuntuOneMacaroons
+
+    @property
+    def value(self) -> UbuntuOneMacaroons:
+        """Get the actual "payload" UbuntuOneMacaroons."""
+        return self.v
+
+    def marshal(self) -> str:
+        """Serialize this Ubuntu One model into a string suitable for storage."""
+        return json.dumps(self.dict())
+
+    @classmethod
+    def unmarshal(cls, marshalled_creds: str) -> "UbuntuOneModel":
+        """Deserialize an Ubuntu One model from previously-stored credentials.
+
+        :param marshalled_creds:
+            The previously-stored Ubuntu One credentials, which can either be in the "new",
+            type-based format or in the "old", dict-based one.
+        """
+        creds = json.loads(marshalled_creds)
+
+        if "t" in creds:
+            return UbuntuOneModel(**creds)
```

Comment:
Can we have UbuntuOne credentials with `t` other than `u1-macaroon`? In this case it would fail to validate the `Literal` type. On the other hand, if it is `u1-macaroon`, then the test would be unnecessary? (It could just be `return cls(**creds)`)

To be completely consistent with the pydantic unmarshal pattern we've been using elsewhere, the `unmarshal` classmethod should already receive deserialized data so maybe we could have the json decoding done in `unmarshal_u1_credentials`?

Reply:
I think there are two points to address here:

> Can we have UbuntuOne credentials with t other than u1-macaroon? In this case it would fail to validate the Literal type. On the other hand, if it is u1-macaroon, then the test would be unnecessary? (It could just be return cls(**creds))

You can have UbuntuOne credentials that are a dict, but without the "t" - those are the current (without this PR) ubuntu one stored creds. See for example the `old_credentials()` fixture in `test_ubuntu_one_store_client.py`. So the `if "t" in creds:` check is really a "if this is a new-style creds payload" check. I should add a comment about that, but do you see a different way of supporting the two cases?

> To be completely consistent with the pydantic unmarshal pattern we've been using elsewhere, the unmarshal classmethod should already receive deserialized data so maybe we could have the json decoding done in unmarshal_u1_credentials?

I see. So the suggested change is to have the `(un)marshal()` calls deal with models and dicts, and the `(un)marsha_u1/candid_creds()` do the json wrapping/unwrapping, right?

Reply:
Updated the code addressing both points:
- Followed conventions in marshal()/unmarshal();
- Improved docs to make the purpose of the code clear, detailing the multiple backwards-comp cases that must be handled by each credential type.